src/AggregateFunctions/AggregateFunctionRangeSum.h

```diff
@@ -0,0 +1,189 @@
+
+namespace DB
+{
+
+template <typename T>
+struct AggregateFunctionRangeSumData
+{
+    using TimeGap = std::pair<T, T>;
+    using TimeGaps = PODArrayWithStackMemory<TimeGap, 64>;
+
+    bool sorted = false;
+
+    TimeGaps gap_list;
+
+    size_t size() const { return gap_list.size(); }
+
+    void add(T start, T end)
+    {
+        if (sorted && gap_list.size() > 0)
+        {
+            sorted = gap_list.back().first <= start;
+        }
+        gap_list.emplace_back(start, end);
+    }
+
+    void merge(const AggregateFunctionRangeSumData & other)
+    {
+        if (other.gap_list.empty())
+            return;
+
+        const auto size = gap_list.size();
+
+        gap_list.insert(std::begin(other.gap_list), std::end(other.gap_list));
+
+        /// either sort whole container or do so partially merging ranges afterwards
+        if (!sorted && !other.sorted)
+            std::stable_sort(std::begin(gap_list), std::end(gap_list));
+        else
+        {
+            const auto begin = std::begin(gap_list);
+            const auto middle = std::next(begin, size);
+            const auto end = std::end(gap_list);
+
+            if (!sorted)
+                std::stable_sort(begin, middle);
+
+            if (!other.sorted)
+                std::stable_sort(middle, end);
+
+            std::inplace_merge(begin, middle, end);
+        }
+
+        sorted = true;
+    }
+
+    void sort()
+    {
+        if (!sorted)
+        {
+            std::stable_sort(std::begin(gap_list), std::end(gap_list));
+            sorted = true;
+        }
+    }
+
+    void serialize(WriteBuffer & buf) const
+    {
+        writeBinary(sorted, buf);
+        writeBinary(gap_list.size(), buf);
+
+        for (const auto & time_gap : gap_list)
+        {
+            writeBinary(time_gap.first, buf);
+            writeBinary(time_gap.second, buf);
+        }
+    }
+
+    void deserialize(ReadBuffer & buf)
+    {
+        readBinary(sorted, buf);
+
+        size_t size;
+        readBinary(size, buf);
+
+        gap_list.clear();
+        gap_list.reserve(size);
+
+        T start, end;
+
+        for (size_t i = 0; i < size; ++i)
+        {
+            readBinary(start, buf);
+            readBinary(end, buf);
+            gap_list.emplace_back(start, end);
+        }
+    }
+};
+
+template <typename T, typename Data>
+class AggregateFunctionRangeSum final : public IAggregateFunctionDataHelper<Data, AggregateFunctionRangeSum<T, Data>>
+{
+private:
+    UInt64 getRangeSum(Data & data) const
+    {
+        if (data.size() == 0)
+            return 0;
+
+        data.sort();
+
+        typename Data::TimeGaps merged;
+
+        for (const auto & gap : data.gap_list)
+        {
+            if (!merged.size() || merged.back().second < gap.first)
+                merged.emplace_back(gap.first, gap.second);
+            else
+                merged.back().second = std::max(merged.back().second, gap.second);
+        }
```

Comment:
I suppose sum can be calculated on one pass through `gap_list` without copying into `merged`, you can try to rewrite this piece or we can do it later.

Reply:
OK. I will rewrite it.